### ISSUE TITLE
Enhancement: `no-unnecessary-type-assertion` detect unnecessary widening assertions

### ISSUE THREAD

**Issue by xonaman** [typescript-eslint/typescript-eslint#12162]
### Description

`no-unnecessary-type-assertion` currently only flags assertions where the pre-assertion and post-assertion types are **referentially identical** (`uncast === cast`). It does not detect assertions that **widen** to a supertype — i.e., where the original type is already assignable to the asserted type.

Meanwhile, `no-unsafe-type-assertion` explicitly permits widening assertions because they are "safe." This creates a gap: **unnecessary widening assertions** are caught by neither rule.

### Example

```ts
interface CountResponse {
  count: number;
  _shards: { total: number; successful: number; failed: number };
  terminated_early?: boolean;
}

declare const res: CountResponse;

// ❌ Should be flagged — CountResponse is already assignable to { count: number }
const count = (res as { count: number }).count;

// ✅ Correct — access the property directly
const count = res.count;
```

Another common pattern this would catch:

```ts
interface User {
  id: string;
  name: string;
  email: string;
}

declare const user: User;

// ❌ Should be flagged — User is already assignable to { id: string }
const id = (user as { id: string }).id;
```

### Proposed Change

Add a `checkWideningAssertions` option (default: `false`) that uses `checker.isTypeAssignableTo(uncastType, castType)` instead of reference equality to detect unnecessary assertions.

Rough implementation sketch in the rule's `TSAsExpression, TSTypeAssertion` visitor:

```ts
if (options.checkWideningAssertions) {
  const uncastType = services.getTypeAtLocation(node.expression);
  const castType = services.getTypeAtLocation(node);

  if (
    checker.isTypeAssignableTo(uncastType, castType) &&
    uncastType !== castType && // already handled by existing logic
    !isTypeFlagSet(uncastType, ts.TypeFlags.Any | ts.TypeFlags.Never)
  ) {
    context.report({
      node,
      messageId: 'unnecessaryWideningAssertion',
    });
  }
}
```

**Edge cases to exclude:**
- `any` and `never` — assignable to everything, but casts from them are intentional
- Identical types — already reported by existing logic
- `as const` — already handled separately

### Why

Widening assertions add visual clutter without changing runtime behavior or type safety. They often indicate a misunderstanding of the original type (the developer didn't realize the property was already available) or are left over from refactoring when the original type was previously `unknown` or untyped.

This is distinct from `no-unsafe-type-assertion` which correctly permits widening because it's "safe" — this proposal is about flagging widening that is **both safe and unnecessary**.

**Comment by ulrichstark** [comment]
Related: #8107 and #11789

**Comment by JoshuaKGoldberg** [comment]
Agreed. #8107 may have a specific title but its discussion agreed on widening its scope to what this issue suggests. I renamed that one just now and am marking this as a dupe. Thanks for filing!
